bundles/org.smarthomej.binding.tcpudp/src/main/java/org/smarthomej/binding/tcpudp/internal/TcpUdpThingHandler.java

```diff
@@ -0,0 +1,348 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                logger.debug("Failed sending '{}' in thing '{}': {}", command, thing.getUID(), e.getMessage());
+            }
+        });
+    }
+
+    protected Optional<ContentWrapper> doTcpSyncRequest(String request) {
+        try (Socket socket = new Socket(config.host, config.port);
+                OutputStream out = socket.getOutputStream();
+                InputStream in = socket.getInputStream();
+                ByteArrayOutputStream outputByteArrayStream = new ByteArrayOutputStream()) {
+            socket.setSoTimeout(config.timeout);
+            out.write(request.getBytes(Objects.requireNonNullElse(config.encoding, StandardCharsets.UTF_8.name())));
+            out.flush();
+            byte[] buffer = new byte[config.bufferSize];
+            int len;
+
+            do {
+                len = in.read(buffer);
+                if (len != -1) {
+                    outputByteArrayStream.write(buffer, 0, len);
+                }
+            } while (len == config.bufferSize);
+            outputByteArrayStream.flush();
+
+            ContentWrapper contentWrapper = new ContentWrapper(outputByteArrayStream.toByteArray(),
+                    Objects.requireNonNullElse(config.encoding, StandardCharsets.UTF_8.name()), null);
+
+            updateStatus(ThingStatus.ONLINE);
+            return Optional.of(contentWrapper);
+        } catch (Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.debug("Failed to request '{}' in thing '{}': {}", request, thing.getUID(), e.getMessage());
+        }
+
+        return Optional.empty();
+    }
+
+    protected void doUdpAsyncSend(String command) {
+        scheduler.execute(() -> {
+            try (DatagramSocket socket = new DatagramSocket()) {
+                socket.setSoTimeout(config.timeout);
+                InetAddress inetAddress = InetAddress.getByName(config.host);
+
+                byte[] buffer = command.getBytes(StandardCharsets.UTF_8);
+                DatagramPacket packet = new DatagramPacket(buffer, buffer.length, inetAddress, config.port);
+
+                socket.send(packet);
+                updateStatus(ThingStatus.ONLINE);
+            } catch (IOException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                logger.debug("Failed sending '{}' in thing '{}': {}", command, thing.getUID(), e.getMessage());
+            }
+        });
+    }
+
+    protected Optional<ContentWrapper> doUdpSyncRequest(String request) {
+        try (DatagramSocket socket = new DatagramSocket()) {
+            socket.setSoTimeout(config.timeout);
+            InetAddress inetAddress = InetAddress.getByName(config.host);
+
+            byte[] buffer = request.getBytes(StandardCharsets.UTF_8);
```

Comment:
Does makes sense to allow to customize this encode as you do with the response encode? Not sure about it but I think it can make sense in some scenarios.
I just saw that in the doTcpSyncRequest you use the config.encode also when getting the bytes to write so, did you forget to do it here or is it intentional?

Reply:
indeed, forgotten